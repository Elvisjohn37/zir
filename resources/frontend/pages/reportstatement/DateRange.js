import React from 'react';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Skeleton from 'frontend/components/Skeleton';
import Button from '@mui/material/Button';
import DateRangeIcon from '@mui/icons-material/DateRange';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import ListItemText from '@mui/material/ListItemText';
import { DateRange } from 'frontend/components/ReportCommonFormatter';
import styles from './DateRange.module';

export function DesktopDateRange({ dateRange, onChange, activeDate }) {
	return (
		<>
			{dateRange.isLoading ? (
				<Skeleton count={3} height={20} className={styles.desktopDateRangeLoader} />
			) : (
				<Tabs
					value={activeDate}
					onChange={onChange}
					textColor="secondary"
					indicatorColor="primary"
					variant="fullWidth"
					aria-label="full width tabs example"
					classes={{ indicator: styles.dateTabIndicator }}
				>
					{dateRange.range.map((row) => (
						<Tab
							classes={{ root: styles.dateButton }}
							component="a"
							value={row.number}
							key={row.number}
							// label={row.dateRange}
							label={<DateRange row={row} />}
							color="secondary"
							disableRipple
						/>
					))}
				</Tabs>
			)}
		</>
	);
}

export function MobileDateRange({ dateRange, activeDate, onChange }) {
	let [anchorEl, setAnchorEl] = React.useState(null);

	let handleClick = (event) => {
		setAnchorEl(event.currentTarget);
	};

	let handleClose = () => {
		setAnchorEl(null);
	};

	let activeDateRangeRow = dateRange.range.find((item) => item.number === activeDate);

	return (
		<>
			{dateRange.isLoading ? (
				<Skeleton count={1} height={32} className={styles.mobileDateRangeLoader} />
			) : (
				<>
					<Button
						variant="contained"
						color="primary"
						disableElevation
						className={styles.mobileDateRangeButton}
						onClick={handleClick}
						endIcon={<DateRangeIcon />}
						classes={{ endIcon: styles.mobileDateRangeIcon }}
					>
						<DateRange row={activeDateRangeRow} />
					</Button>
					<Menu anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={handleClose}>
						{dateRange.range.map((row, index) => (
							<MenuItem
								key={index}
								selected={activeDate == row.number}
								onClick={(e) => {
									onChange(e, row.number);
									handleClose();
								}}
								component="a"
							>
								<ListItemText primary={<DateRange row={row} />} />
							</MenuItem>
						))}
					</Menu>
				</>
			)}
		</>
	);
}
