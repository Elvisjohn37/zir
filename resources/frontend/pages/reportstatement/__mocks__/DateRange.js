import React from 'react';
import { componentModuleMock, transformProps } from 'frontend/tests/helpers';
let mockedModules = componentModuleMock(jest.createMockFromModule('../DateRange'));

mockedModules.DesktopDateRange = ({ children, ...props }) => {
	let { newProps } = transformProps(props);
	return (
		<>
			{props.dateRange.isLoading ? (
				<div {...newProps} data-testid="DesktopDateRange">
					Skeleton Loader
				</div>
			) : (
				<div {...newProps} data-testid="DesktopDateRange">
					<span>activeDate: {props.activeDate}</span>
					<span>row: {JSON.stringify(props.row)}</span>
				</div>
			)}
		</>
	);
};

mockedModules.MobileDateRange = ({ children, ...props }) => {
	let { newProps } = transformProps(props);
	return (
		<>
			{props.dateRange.isLoading ? (
				<div {...newProps} data-testid="MobileDateRange">
					Skeleton Loader
				</div>
			) : (
				<div {...newProps} data-testid="MobileDateRange">
					<span>activeDate: {props.activeDate}</span>
					<span>row: {JSON.stringify(props.row)}</span>
				</div>
			)}
		</>
	);
};

module.exports = mockedModules;
