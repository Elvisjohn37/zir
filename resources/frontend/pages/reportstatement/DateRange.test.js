import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DesktopDateRange, MobileDateRange } from './DateRange';

jest.mock('frontend/components/Skeleton');
jest.mock('frontend/components/ReportCommonFormatter');
jest.mock('@mui/material/Tabs');
jest.mock('@mui/material/Tab');
jest.mock('@mui/material/Button');
jest.mock('@mui/material/Menu');
jest.mock('@mui/material/MenuItem');
jest.mock('@mui/material/ListItemText');
jest.mock('@mui/icons-material/DateRange');

const DATE_RANGE_ISLOADING = { isLoading: true, range: [] };
const DATE_RANGE_LOADED = {
	isLoading: false,
	range: [
		{ number: 3, startDate: '2021-07-01', endDate: '2021-07-31' },
		{ number: 2, startDate: '2021-08-01', endDate: '2021-08-31' },
		{ number: 1, startDate: '2021-09-01', endDate: '2021-09-01' },
	],
};

describe('</DesktopDateRange>', () => {
	let defaultProps = { dateRange: DATE_RANGE_LOADED, onChange: () => jest.fn() };
	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
		cleanup();
	});

	it('should match snapshot', () => {
		let { asFragment } = render(<DesktopDateRange {...defaultProps} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should display loader when config is still loading', () => {
		render(<DesktopDateRange {...defaultProps} dateRange={DATE_RANGE_ISLOADING} />);
		expect(screen.getByTestId('Skeleton')).toBeInTheDocument();
		expect(screen.queryByTestId('Tabs')).not.toBeInTheDocument();
	});

	it('should display date range options in tab form after config loading', () => {
		render(<DesktopDateRange {...defaultProps} />);
		expect(screen.queryByTestId('Skeleton')).not.toBeInTheDocument();
		DATE_RANGE_LOADED.range.map((item) => {
			expect(screen.getByText('{"row":' + JSON.stringify(item) + '}')).toBeInTheDocument();
		});
	});

	it('should handle change when user click an option and highlight active date', () => {
		let handleChange = jest.fn();
		render(<DesktopDateRange {...defaultProps} onChange={handleChange} />);
		userEvent.click(screen.getByTestId('Tabs'));
		expect(handleChange).toHaveBeenCalled();
	});
});

describe('</MobileDateRange>', () => {
	let defaultProps = { activeDate: 1, dateRange: DATE_RANGE_LOADED, onChange: () => jest.fn() };
	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
		cleanup();
	});

	it('should match snapshot', () => {
		let { asFragment } = render(<MobileDateRange {...defaultProps} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should display loader when config is still loading', () => {
		render(<MobileDateRange {...defaultProps} dateRange={DATE_RANGE_ISLOADING} />);
		expect(screen.getByTestId('Skeleton')).toBeInTheDocument();
		expect(screen.queryByTestId('Tabs')).not.toBeInTheDocument();
	});

	it('should display date range options in button menu after config loading', () => {
		render(<MobileDateRange {...defaultProps} />);
		expect(screen.queryByTestId('Skeleton')).not.toBeInTheDocument();
		expect(screen.queryByTestId('Button')).toBeInTheDocument();
	});

	it('should handle change when user click an option and highlight active date', () => {
		let handleChange = jest.fn();
		render(<MobileDateRange {...defaultProps} onChange={handleChange} />);
		fireEvent.change(screen.getByTestId('Button'));

		DATE_RANGE_LOADED.range.map((item) => {
			let activeCounter = 0;
			let inActiveCounter = 0;
			if (item.number == 1) {
				expect(screen.getAllByText('selected: true')[activeCounter]).toBeInTheDocument();
				activeCounter++;
			} else {
				expect(screen.getAllByText('selected: false')[inActiveCounter]).toBeInTheDocument();
				inActiveCounter++;
			}
		});
	});
});
