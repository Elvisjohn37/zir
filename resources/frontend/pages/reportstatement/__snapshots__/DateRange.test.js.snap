// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`</DesktopDateRange> should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="full width tabs example"
    classes="{\\"indicator\\":\\"dateTabIndicator\\"}"
    data-testid="Tabs"
    indicatorcolor="primary"
    textcolor="secondary"
    variant="fullWidth"
  >
    <a
      classes="{\\"root\\":\\"dateButton\\"}"
      color="secondary"
      data-testid="Tab"
      disableripple="true"
      value="3"
    >
      <div>
        {"row":{"number":3,"startDate":"2021-07-01","endDate":"2021-07-31"}}
      </div>
    </a>
    <a
      classes="{\\"root\\":\\"dateButton\\"}"
      color="secondary"
      data-testid="Tab"
      disableripple="true"
      value="2"
    >
      <div>
        {"row":{"number":2,"startDate":"2021-08-01","endDate":"2021-08-31"}}
      </div>
    </a>
    <a
      classes="{\\"root\\":\\"dateButton\\"}"
      color="secondary"
      data-testid="Tab"
      disableripple="true"
      value="1"
    >
      <div>
        {"row":{"number":1,"startDate":"2021-09-01","endDate":"2021-09-01"}}
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`</MobileDateRange> should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="mobileDateRangeButton"
    classes="{\\"endIcon\\":\\"mobileDateRangeIcon\\"}"
    color="primary"
    data-testid="Button"
    disableelevation="true"
    variant="contained"
  >
    <div
      data-testid="DateRange"
    >
      2021-09-01 - 2021-09-01
    </div>
    <div>
      {}
    </div>
  </button>
  <div
    data-testid="Menu"
    keepmounted="true"
  >
    <div
      component="a"
      data-testid="MenuItem"
    >
      selected: false
      <div
        data-testid="ListItemText"
      >
        <div>
          {"row":{"number":3,"startDate":"2021-07-01","endDate":"2021-07-31"}}
        </div>
      </div>
    </div>
    <div
      component="a"
      data-testid="MenuItem"
    >
      selected: false
      <div
        data-testid="ListItemText"
      >
        <div>
          {"row":{"number":2,"startDate":"2021-08-01","endDate":"2021-08-31"}}
        </div>
      </div>
    </div>
    <div
      component="a"
      data-testid="MenuItem"
    >
      selected: true
      <div
        data-testid="ListItemText"
      >
        <div>
          {"row":{"number":1,"startDate":"2021-09-01","endDate":"2021-09-01"}}
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
